### PR TITLE
Flash Messages Application's recurrency not working #10

### PR DESCRIPTION
**Pull request by oanalavinia**
I've added in the validation class another property that refers to the date when the message was last seen so I can display it if that date does not correspond with today

### CODE REVIEW

File: application-flashmessages-ui/src/main/resources/Flash/DisplayFlashMessages.xml [hunk 1]
```diff
@@ -154,8 +154,15 @@
   #set($msgDoc = $xwiki.getDocument($msg.doc))
   #set($msgValObj = $msgDoc.getObject('Flash.FlashValidationClass'))
   #set($msgValList = $msgValObj.getProperty('flash').getValue())
+  #set($msgDateSeen = $msgValObj.getProperty('seenDate').getValue())
```
Comment (acotiuga) [inline application-flashmessages-ui/src/main/resources/Flash/DisplayFlashMessages.xml]: You can also use the shorter form `$msgValObj.getValue('seenDate')`.

File: application-flashmessages-ui/src/main/resources/Flash/DisplayFlashMessages.xml [hunk 2]
```diff
@@ -154,8 +154,15 @@
...
+  #set($dateToday = $datetool.format('yyyy-M-d',$dateNow))
```
Comment (acotiuga) [inline application-flashmessages-ui/src/main/resources/Flash/DisplayFlashMessages.xml]: `yyyy-M-d` doesn't look like the default date format (`dd/MM/yyyy HH:mm:ss` is used when creating an XClass with a Date property). Not sure why did you choose the other one.

File: application-flashmessages-ui/src/main/resources/Flash/DisplayFlashMessages.xml [hunk 3]
```diff
@@ -154,8 +154,15 @@
...
+  #set($seenDate = $datetool.format('yyyy-M-d',$msgDateSeen))
+  #set($seenToday = true)
+  #if($dateToday != $seenDate)
+    #set($seenToday = false)
+  #end
```
Comment (acotiuga) [inline application-flashmessages-ui/src/main/resources/Flash/DisplayFlashMessages.xml]: Shorter form `#set ($seenToday = $dateToday == $seenDate)`.

File: application-flashmessages-ui/src/main/resources/Flash/DisplayFlashMessages.xml [hunk 4]
```diff
@@ -154,8 +154,15 @@
...
   ##
-  #if(!$userValList.contains($msg.doc) &amp;&amp; !$msgValList.contains($xcontext.user))
+  #if(!$userValList.contains($msg.doc) &amp;&amp; !$msgValList.contains($xcontext.user) || ($seenToday == false))
```
Comment (acotiuga) [inline application-flashmessages-ui/src/main/resources/Flash/DisplayFlashMessages.xml]: You don't really need  extra `()` for the last condition.

File: application-flashmessages-ui/src/main/resources/Flash/FlashNotifier.xml [hunk 5]
```diff
@@ -59,13 +60,24 @@
       #set($list = $obj.getProperty('flash').getValue())
       #if($list == $util.getNull())
         #set($discard = $obj.set('flash', [$xcontext.user]))
+        #set($discard = $obj.set('seenDate', $dateNow))
         #set($discard = $msgdoc.saveAsAuthor("$xcontext.user has seen the message"))
       #elseif($list.class.name == 'com.xpn.xwiki.objects.ListProperty$NotifyList' &amp;&amp; !$list.contains($xcontext.user))
         #set($discard = $list.add($xcontext.user))
         #set($discard = $obj.set('flash', $list))
+        #set($discard = $obj.set('seenDate', $dateNow))
```
Comment (acotiuga) [inline application-flashmessages-ui/src/main/resources/Flash/FlashNotifier.xml]: This line is duplicated few lines above. You can set the `seenDate` property outside the `if-else` statement. I see that the same happens with `#set($discard = $msgdoc.saveAsAuthor("$xcontext.user has seen the message"))`.

File: application-flashmessages-ui/src/main/resources/Flash/FlashNotifier.xml [hunk 6]
```diff
@@ -59,13 +60,24 @@
...
         #set($discard = $msgdoc.saveAsAuthor("$xcontext.user has seen the message"))
       #end
     #end
+    ## Check if the message should be mark as seen today
+    #set($msgObj = $msgdoc.getObject('Flash.FlashValidationClass'))
+    #set($msgDateSeen = $msgObj.getProperty('seenDate').getValue())
+    #set($dateToday = $datetool.format('yyyy-M-d',$dateNow))
+    #set($seenDate = $datetool.format('yyyy-M-d',$msgDateSeen))
```
Comment (acotiuga) [inline application-flashmessages-ui/src/main/resources/Flash/FlashNotifier.xml]: Same as above about the date format. You can even check the default date format of the wiki. See: https://github.com/xwiki/xwiki-platform/commit/495e81f881e7e7d4886ba075974f607207500313#diff-b63753a40cda0f6a848b4d11bee1b84eR219.

File: application-flashmessages-ui/src/main/resources/Flash/FlashNotifier.xml [hunk 7]
```diff
@@ -59,13 +60,24 @@
...
+    #if($dateToday != $seenDate)
+      #set($discard = $msgObj.set('seenDate', $dateNow))
+      #set($discard = $msgdoc.saveAsAuthor("$xcontext.user has seen the message"))
```
Comment (acotiuga) [inline application-flashmessages-ui/src/main/resources/Flash/FlashNotifier.xml]: I see again duplicated code, maybe there's something to be improved here.

File: application-flashmessages-ui/src/main/resources/Flash/DisplayFlashMessages.xml (hunk 2 above)
Comment (oanalavinia) [reply application-flashmessages-ui/src/main/resources/Flash/DisplayFlashMessages.xml]: I just needed to get the day to see if it's different. So `dd/MM/yyyy` would be better? Or I should also consider the time?

Comment (acotiuga) [reply application-flashmessages-ui/src/main/resources/Flash/DisplayFlashMessages.xml]: Sure, take only what you need.

Comment (oanalavinia) [reply application-flashmessages-ui/src/main/resources/Flash/DisplayFlashMessages.xml]: Ok, thanks

**Comment by acotiuga** [comment]
Why is this open if it was merged?

**Comment by oanalavinia** [comment]
> Why is this open if it was merged?

The merged PR contains only the fix of the specified issue, while on this PR some tests were also updated. This can be merged once docker tests will be supported for paid apps.

**Comment by mflorea** [comment]
@oanalavinia if this issue was fixed but we're missing the functional tests then I would create another issue for the functional tests and close this one. WDYT?

**Comment by oanalavinia** [comment]
The issue is already closed, only this PR remained open. But I will create another issue only for functional tests
I'll close this PR and open another one on the specific issue when docker tests will be supported.

**Pull request closed without merging**